pkg/clusteragent/autoscaling/controller.go

```diff
@@ -0,0 +1,152 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoscaling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Controller is generic controller that can be used to implement a controller for a specific resource
+type Controller struct {
+	processor Processor
+	synced    cache.InformerSynced
+	context   context.Context
+
+	// Fields available to child controllers
+	ID        string
+	Client    dynamic.Interface
+	Lister    cache.GenericLister
+	Workqueue workqueue.RateLimitingInterface
+	IsLeader  func() bool
+}
+
+// NewController returns a new workload autoscaling controller
+func NewController(
+	controllerID string,
+	processor Processor,
+	client dynamic.Interface,
+	informer dynamicinformer.DynamicSharedInformerFactory,
+	gvr schema.GroupVersionResource,
+	isLeader func() bool,
+	observable Observable,
+) (*Controller, error) {
+	mainInformer := informer.ForResource(gvr)
+	c := &Controller{
+		processor: processor,
+		ID:        controllerID,
+		Client:    client,
+		Lister:    mainInformer.Lister(),
+		synced:    mainInformer.Informer().HasSynced,
```

Comment:
Is there a reason why we don't use the `HasSynced` method of the EventHandler ? From the public doc of the function:

>// HasSynced returns true if the shared informer's store has been
	// informed by at least one full LIST of the authoritative state
	// of the informer's object collection.  This is unrelated to "resync".
	//
	// Note that this doesn't tell you if an individual handler is synced!!
	// For that, please call HasSynced on the handle returned by
	// AddEventHandler.